sound/soc/sof/intel/byt.c

```diff
@@ -160,13 +160,31 @@ static void byt_dump(struct snd_sof_dev *sdev, u32 flags)
 static irqreturn_t byt_irq_handler(int irq, void *context)
 {
 	struct snd_sof_dev *sdev = context;
-	u64 isr;
+	u64 ipcx, ipcd;
 	int ret = IRQ_NONE;
 
-	/* Interrupt arrived, check src */
-	isr = snd_sof_dsp_read64(sdev, BYT_DSP_BAR, SHIM_ISRX);
-	if (isr & (SHIM_ISRX_DONE | SHIM_ISRX_BUSY))
+	ipcx = snd_sof_dsp_read64(sdev, BYT_DSP_BAR, SHIM_IPCX);
+	ipcd = snd_sof_dsp_read64(sdev, BYT_DSP_BAR, SHIM_IPCD);
+
+	if (ipcx & SHIM_BYT_IPCX_DONE) {
+
+		/* reply message from DSP, Mask Done interrupt first */
+		snd_sof_dsp_update_bits64_unlocked(sdev, BYT_DSP_BAR,
+						   SHIM_IMRX,
+						   SHIM_IMRX_DONE,
+						   SHIM_IMRX_DONE);
```

Comment:
If you don't check the SHIM_ISRX here, we could mask the SHIM_IMRX_DONE multiple times if a new message from DSP arrives and the previous reply message from DSP has not been thread handled yet, though this could be not harmful.

Reply:
SHIM_ISRX_DONE is set when LPE writes 1 in IPCX_DONE, so this is redundant/mirrorred. You cannot have different values for SHIM_ISRX_DONE and IPCX_DONE.

Reply:
I believe we actually could, imagine this:
1. a new host -> LPE IPC A;
2. LPE handled A, and send reply for A to host. This will set ISRX_DONE and IPCX_DONE;
3. Host get ISRX_DONE interrupt, unmask IMRX_DONE here and waking up the IRQ tread handler, but at this time, if
4. LPE want to send a new IPC (e.g. tracing position update IPC) to the host, it will set ISRX_BUSY and IPCD_BUSY;
5. Host get ISRX_BUSY interrupt, we will enter byt_irq_handler() here again but the clearing of IPCX_DONE after step 3 above (doing in the thread byt_irq_thread) may not be done yet, in this scenario, the check of "if (ipcx & SHIM_BYT_IPCX_DONE)" will be true and we will mask the SHIM_IMRX_DONE again.

Reply:
we never clear SHIM_ISRX_DONE and masking does not affect the values of IPXC_DONE registers, I don't understand your point *at all*.
Is anything broken?

Reply:
If you add some print log as I did in https://github.com/thesofproject/linux/issues/1492, and compile FW with verbose trace enabled, then running a stream, and you could see that if what I described above will happen.

Reply:
> we never clear SHIM_ISRX_DONE and masking does not affect the values of IPXC_DONE registers, I don't understand your point _at all_.
> Is anything broken?

As I commented at the beginning, this could be not harmful so nothing should be broken by it.

Reply:
@keyonjie adding a test for ISRX_DONE would not change anything in step 5, the bit would be set just like IMRX_DONE, so in this case there is no way to prevent the IMRX_DONE bit from being set.